### PR TITLE
Allow nested reverse mapped type printback

### LINKED ISSUE #40003: d.ts type is different than intellisense type
**TypeScript Version:**  3.9.6 (npm and in vs code)

**Search Terms:**
declaration intellisense

**Code**

```
export type Validator<T> = NativeTypeValidator<T> | ObjectValidator<T>

export type NativeTypeValidator<T> = (n: any) => T | undefined
export type ObjectValidator<O> = {
  [K in keyof O]: Validator<O[K]> 
}

//native validators
export const SimpleStringValidator:NativeTypeValidator<string> = (val) => typeof(val) === "string" ? val : undefined

///object validator function
export const ObjValidator = <V>(validatorObj: ObjectValidator<V>) => (o:any):V =>{
  let result = {} as V;
  //we can only validate objects
  if (typeof (o) !== "object") { return undefined; }
  const validatorKeys = Object.keys(o) as [keyof ObjectValidator<V>]
  validatorKeys.forEach((validatorKey) => { 
    const objValue = o[validatorKey] as V[keyof V];
    const objectValidator = validatorObj[validatorKey]
    if (!objectValidator) { return undefined } //do nothing if no validator exists for the key in o
    //figure out if we have a nested object validator or a native validator at the corresponding key of validatorObj
    if (typeof (objectValidator) === "object") {
      result[validatorKey] = ObjValidator(objectValidator as ObjectValidator<V[keyof V]>)(objValue)
    }
    else {
      const nativeValidator = objectValidator as NativeTypeValidator<V[keyof V]>;
      result[validatorKey] = nativeValidator(objValue)
    }
  })
  return result;
}

export const test  = {
  Test: {
    Test1: {
      Test2: SimpleStringValidator
    },
  }
}

export const validatorFunc = ObjValidator(test);
export const outputExample = validatorFunc({
  Test: {
    Test1: {
      Test2: "hi"
    },
  }
})

outputExample.Test.Test1.Test2 = "1";
outputExample.Test.Test1.Test2 = 1; //vs code intellisense complains because needs to be type string

```

**Expected behavior:**
I'm trying to extract a type mapping from an object containing native type validation function so for example, if I have the object 

```
const test  = {
  Test: {
    Test1: {
      Test2: SimpleStringValidator //return type is string or undefined but input can be anything
    },
  }
}
```

I want to generate the type.
```
type Extracted = {
  Test: {
    Test1: {
      Test2: string
    }
  }
}
```
and have it be the same in intellisense and in the generated type declaration file. In the above example this generated type should be the return value type of the 

**Actual behavior:**
Intellisense and the output declaration file are inconsistent. Intellisense creates a nested type with type string at property Test2 (this is the desired behaviour) but the output declaration file is as follows 

```
export declare type Validator<T> = NativeTypeValidator<T> | ObjectValidator<T>;
export declare type NativeTypeValidator<T> = (n: any) => T | undefined;
export declare type ObjectValidator<O> = {
    [K in keyof O]: Validator<O[K]>;
};
export declare const SimpleStringValidator: NativeTypeValidator<string>;
export declare const ObjValidator: <V>(validatorObj: ObjectValidator<V>) => (o: any) => V;
export declare const test: {
    Test: {
        Test1: {
            Test2: NativeTypeValidator<string>;
        };
    };
};
export declare const validatorFunc: (o: any) => {
    Test: {
        Test1: any;
    };
};
export declare const outputExample: {
    Test: {
        Test1: any;
    };
};
```

In the declaration file the Test1 definition gets blown away to any, which leads me to not be able to directly import the generated type into other projects. 

**Playground Link:** 
The issue is reproducible in the typescript playground. Intellisense displays the Test2 property correctly as type string, where in the d.ts file Test1 is of type any. Tried with typescript versions 3.9.2 and 4.0.0-beta
[Playground Link](https://www.typescriptlang.org/play?strictNullChecks=false&experimentalDecorators=false&target=6&jsx=0&ts=4.0.0-beta#code/FAUwHgDg9gTgLgAjgTwiBA1AhgGwJYAmWcsAPACoB8CAvAgHLF4BuI5qI2+RJMF1AHwQB5AEYArEAGM4XQsTJVgoSLEQo0DJq3Zo5PRdToAKAHYAuBFlPIAlLWrkEQgK6mCIAGZ5TIAiuh4JA4RCWlZXHleUmEjBABvYAQEAG0AaQQfBABrEGQoTxEAXUt9BT5hdKLqYABfZQB6BtNtdGZIgxgAZwC1BCkoUy7EAGU8AFsIHBARuBgfAHMy3nNGOBY2DmWyYfnTBbjjdpx7GmoNEAKj3FOaOgAiXcX7hAB+BGOESzcPb19-YBNBBNKBhGQfDrlBCeNwyPCDXpBAZDRBicTbGC0BCkDCUa7ccpoyxo8IYnGUU7UYxQczWOzmDAORLJaaIGAgLouHCIOjxWpWLqYADcSWBDQA7ugpNYEIMcMgIQS4OhQZIZD1knhCsYLgUENT7ABCO4Ie6q8L3ezxBDsuAuGCmBA-Lw+PxChD1ZLI4aKqKwNJ5QV0EkyAB0uWQXQNAtSEb1IYiSrIuKKouOfpgAcjoc8sAAolgpAALYz4jNZykJBCir2DH2qrgudB0KApdOdLNFGMYFJxwoYIoi5K1lGysGJjNY9uEiRtyG8Ts1zLaw3mmQYq02kB2h1O9wu-4esUEKAIUxQOBFxbLs+n6e8BDgPDDQW5zGX9ARzKOqBLpreBZ7RVFxEC1BBJQQIssFYKwzw5ZUCDHNVEHvWBZUxLAz1aX1OisdQiylWB2S6aB3GvL89VQmA0SXMCdQ4PVqXHDdaBNM1x0tBIl2SYiuTgOck0zPIu2DCQMSY5CMRjBMyR7PtMGqWwJMbEBbCXT1hxAHAunQZlhxHH0WnWVgpJbZj5zQrBBTWDZdE4Cy+DkvI9QHSgh30rdOW5ATy2ErEjI2cSG1wJs1P0jTajCrcd0dXjuRFepEUQb11Hg5JeVFch4MsPTkiy4YAEYcu4hB8rgAAmSwxkmaZZj2JYHPUgAaUV6kS8BAmSusUIcgAxWEsTRcTlWGWwRQ6voUtlECIBAvMwCwGrmxw8p+tMKRjD0sriv0sqiq4jzSvgyrTSve5mtauo1OAKAZrmhaltDMqnvggqXuGcqsXuAr7hFW64FmuB5sWqYQHeuBwbesrProAr3SaZhBQGDxv2VHB8B0oZCJqrAfEFURpCwFwdLgvxBRIBACeCTQfDgYAgA)

ts compiler options:
```
  "compilerOptions": {
    "noImplicitAny": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2019",
    "module": "ESNext"
  }
```
**Related Issues:**

### PR DESCRIPTION
**Pull request by weswigham**
When the reverse mapped type in question both isn't already being printed, and isn't the child type of a reverse mapping over a non-anonymous type (eg, `Window` or `Number`). This means that reverse mappings over object literal expression types should be printed back in full, rather than getting truncated.

To support this change, I've added caching into reverse mapped symbol type lookup (previously we'd redo the inference every time we asked for the type of the symbol), _and_ introduced a simplification into the construction of reverse mapped symbols so we manufacture less type identities when doing deep reverse mappings (by recognizing that `{[K in keyof T[K_1]]: Whatever<T[K_1]>]}` is equivalent to `{[K in keyof T]: Whatever<T>]}` as far as reverse mapped symbol type inference is concerned). Without that, we'd manufacture new sets of type identities for every type mapped over for every level "down" we had to recur within a deeply-reverse mapped type.

Fixes #40003 in most cases.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot perf test this
@typescript-bot user test this
@typescript-bot test this
@typescript-bot run dt

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5058,9 +5058,21 @@ namespace ts {
...
+            function shouldUsePlaceholderForProperty(propertySymbol: Symbol, context: NodeBuilderContext) {
+                // Use placeholders for reverse mapped types we've either already descended into, or which
+                // are nested reverse mappings within a mapping over a non-anonymous type. The later is a restriction mostly just to
+                // reduce the blowup in printback size from doing, eg, a deep reverse mapping over `Window`.
+                // Since anonymous types usually come from expressions, this allows us to preserve the output
+                // for deep mappings which likely come from expressions, while truncating those parts which
+                // come from mappings over library functions.
+                return !!(getCheckFlags(propertySymbol) & CheckFlags.ReverseMapped) && (
+                    contains(context.reverseMappedStack, propertySymbol as ReverseMappedSymbol)
+                    || (context.reverseMappedStack?.[0] && !(getObjectFlags(last(context.reverseMappedStack).propertyType) & ObjectFlags.Anonymous)));
```
Comment (sandersn) [inline src/compiler/checker.ts:5070]: tiny style suggestion: put the `&&` on the next line to highlight that it's the parent of the `||`.
```suggestion
                return !!(getCheckFlags(propertySymbol) & CheckFlags.ReverseMapped) 
                    && (
                     contains(context.reverseMappedStack, propertySymbol as ReverseMappedSymbol)
                     || (context.reverseMappedStack?.[0] && !(getObjectFlags(last(context.reverseMappedStack).propertyType) & ObjectFlags.Anonymous)));

```

Comment (sandersn) [reply src/compiler/checker.ts:5070]: Also, I don't understand what the last condition means and how it maps back to "nested reverse mapping". Is it any reverse mapping at all? Does `context.reverseMappedStack?.[0]` being defined mean that we're in a nested one?

**Review by sandersn — COMMENTED** [review]
- I like the behaviour change, 
- agree with the index access simplification
- have a question about the nested printback
- don't feel qualified to OK the caching change (perf tests help here although I don't think we test mapped types at all on them)

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts:5070]: Yep. The stack having members means we're within a reverse mapping (essentially the same as the old context flag) - the last member of the stack is just also looked at now, to see if it's a type we should printback anyway.

**Review by ahejlsberg — APPROVED** [review]

**Pull request merged**